### DISCUSSION TITLE
An invalid key plugin_paths was specified

### DISCUSSION THREAD

**Discussion by srchild** [mcguirepr89/BirdNET-Pi#740] [General]
Just updated and update hung for several minutes. Rebooted and seeing this in the log:

11:03:48---Traceback (most recent call last):
11:03:48---  File "/usr/local/bin/server.py", line 16, in <module>
11:03:48---    from utils.notifications import sendAppriseNotifications
11:03:48---  File "BirdNET-Pi/scripts/utils/notifications.py", line 16, in <module>
11:03:48---    asset = apprise.AppriseAsset(
11:03:48---  File "BirdNET-Pi/birdnet/lib/python3.9/site-packages/apprise/AppriseAsset.py", line 157, in __init__
11:03:48---    raise AttributeError(
11:03:48---AttributeError: AppriseAsset init(): An invalid key plugin_paths was specified.

In the code tab I see notifications.py "Merge pull request  https://github.com/mcguirepr89/BirdNET-Pi/pull/724 [from knmurphy/enable-apprise-plugins

That has 

asset = apprise.AppriseAsset(
    plugin_paths=[
        userDir + "/.apprise/plugins",
        userDir + "/.config/apprise/plugins",
    ]
)

I have no .apprise directory on my system. I do have .config but it does not have an apprise subdirectory

**Comment by srchild** [comment]
birdnet/lib64/python3.9/site-packages/apprise/AppriseAsset.py does not initialise plugin_paths

    def __init__(self, **kwargs):
        """
        Asset Initialization

        """
        # Assign default arguments if specified
        for key, value in kwargs.items():
            if not hasattr(AppriseAsset, key):
                raise AttributeError(
                    'AppriseAsset init(): '
                    'An invalid key {} was specified.'.format(key))

            setattr(self, key, value)



In my other installation (which I have not updated today) plugin_paths is initialised in  birdnet/lib64/python3.9/site-packages/apprise/AppriseAsset.py

    def __init__(self, plugin_paths=None, **kwargs):
        """
        Asset Initialization

        """
        # Assign default arguments if specified
        for key, value in kwargs.items():
            if not hasattr(AppriseAsset, key):
                raise AttributeError(
                    'AppriseAsset init(): '
                    'An invalid key {} was specified.'.format(key))

            setattr(self, key, value)

        if plugin_paths:
            # Load any decorated modules if defined
            module_detection(plugin_paths)

**Comment by ehpersonal38** [comment]
I'm sorry about this - I committed some code to the repo which seems to have broken things. I've reverted the change so an update should bring your installation back to a working state. (Although I see you've already reinstalled.)

**Comment by caronc** [comment]
If you don't mind, I'm just observing this, but if there is a way i can add more bulletproofing to Apprise, let me know.

Was the plugin path added pointing back to a large library? I'm trying to understand what it tried to load under the hood in this case

**Reply by ehpersonal38** [reply]
Here's the pull request that was merged in, it has some additional details: https://github.com/mcguirepr89/BirdNET-Pi/pull/724

Other than that I'm not sure on any of the technical details.

**Reply by caronc** [reply]
Shouldn't cause any harm unless the version of Apprise on the system wasn't at least v1.0. Maybe in the PR reference, the `requirements.txt` didn't include `apprise>=1.0.0`?

**Reply by ehpersonal38** [reply]
I did some tests with different Apprise versions and I can confirm your suspicions. Anything < version 1.0 results in the same error as the OP.

I'll implement a fix for this soon - thanks @caronc for troubleshooting this.
